### PR TITLE
Add support to specify resource endpoint for metrics

### PR DESCRIPTION
**Pull request by ms-hujia**
For cloud other than public Azure, a different resource endpoint for MSI retrieving might be used. This PR adds a method to override the default one used when sending metrics.

### CODE REVIEW

File: source/plugins/ruby/out_mdm.rb [hunk 1]
```diff
@@ -119,6 +119,10 @@ def start
             if @isAADMSIAuth && !@isWindows
               @log.info "using IMDS sidecar endpoint for MSI token since its Arc k8s and Linux node"
               @useMsi = true
+              custom_resource_endpoint = ENV["customResourceEndpoint"]
+              if !custom_resource_endpoint.nil? && !custom_resource_endpoint.empty?
+                @@token_resource_audience = custom_resource_endpoint.strip
```
Comment (ganga1980) [inline source/plugins/ruby/out_mdm.rb:124]: what's the custom token resource audience look like? can you please share the audience names?

File: charts/azuremonitor-containers/templates/omsagent-daemonset.yaml [hunk 2]
```diff
@@ -111,6 +111,10 @@ spec:
        - name: CUSTOM_METRICS_ENDPOINT
          value: {{ .Values.omsagent.metricsEndpoint | quote }}
        {{- end }}
+       {{- if ne .Values.omsagent.resourceEndpoint "<your_resource_endpoint>" }}
```
Comment (ganga1980) [inline charts/azuremonitor-containers/templates/omsagent-daemonset.yaml:114]: Please note that, our base branch is ci_prod so you will have to make PRs against ci_prod branch not the ci_dev.

Comment (ms-hujia) [reply charts/azuremonitor-containers/templates/omsagent-daemonset.yaml:114]: Got it, rebased.

File: source/plugins/ruby/out_mdm.rb (hunk 1 above)
Comment (ms-hujia) [reply source/plugins/ruby/out_mdm.rb:124]: For now, we suppose it to be in the format `https://monitor.{FQDN}/` while no final decision has been done yet.

File: charts/azuremonitor-containers/templates/omsagent-daemonset.yaml (hunk 2 above)
Comment (ganga1980) [inline charts/azuremonitor-containers/templates/omsagent-daemonset.yaml:114]: we have changed our agent name from "omsagent" to "ama-logs" so this needs to be updated.

Comment (ganga1980) [inline charts/azuremonitor-containers/templates/omsagent-daemonset.yaml:114]: rename "resourceEndpoint" to "tokenAudience"

File: source/plugins/ruby/out_mdm.rb (hunk 1 above)
Comment (ganga1980) [reply source/plugins/ruby/out_mdm.rb:124]: > For now, we suppose it to be in the format `https://monitor.{FQDN}/` while no final decision has been done yet.

> Did you test this change for both metrics and LA ingestion in Managed Identity mode?

File: charts/azuremonitor-containers/values.yaml [hunk 3]
```diff
@@ -65,6 +65,7 @@ omsagent:
   proxy: <your_proxy_config>
   # This metricsEndpoint used to define the endpoint custom metrics emit to. If not defined, default public Azure monitoring endpoint '{aks_region}.monitoring.azure.com' will be used.
   metricsEndpoint: <your_metrics_endpoint>
+  resourceEndpoint: <your_resource_endpoint>
```
Comment (ganga1980) [inline charts/azuremonitor-containers/values.yaml:68]: For Arc K8s, the audience specified here - https://github.com/microsoft/Docker-Provider/blob/ci_prod/charts/azuremonitor-containers/templates/ama-logs-arc-k8s-crd.yaml#L22 which will indicate the Arc K8s extension platform to get the token with the specified audience.

File: charts/azuremonitor-containers/templates/omsagent-daemonset.yaml (hunk 2 above)
Comment (ms-hujia) [reply charts/azuremonitor-containers/templates/omsagent-daemonset.yaml]: Updated.

Comment (ms-hujia) [reply charts/azuremonitor-containers/templates/omsagent-daemonset.yaml]: Renamed all occurrences in YAML files.

File: source/plugins/ruby/out_mdm.rb (hunk 1 above)
Comment (ms-hujia) [reply source/plugins/ruby/out_mdm.rb:124]: This PR is only intended to support Metrics in MSI mode. Will make changes for Logs in a separate PR later.

File: charts/azuremonitor-containers/values.yaml (hunk 3 above)
Comment (ms-hujia) [reply charts/azuremonitor-containers/values.yaml]: Thanks for the information! Does this mean the corresponding Ruby code will be changed to get IMDS token from K8s CRD later?

Comment (ganga1980) [reply charts/azuremonitor-containers/values.yaml]: Hi, @ms-hujia ,  The audience which we have is fixed. whatever the audience defined in the CRD matches with ruby code.  In your case, if the audience is custom, in that case, ruby code should use the same audience defined in the CRD.   do you have the steps which you have followed to validate this change?

File: source/plugins/ruby/out_mdm.rb (hunk 1 above)
Comment (ganga1980) [reply source/plugins/ruby/out_mdm.rb:124]: > This PR is only intended to support Metrics in MSI mode. Will make changes for Logs in a separate PR later.

> @ms-hujia , are you using the Arc K8s addon-token adapter which we have as side car container? If yes, then audience needs to be updated in the CRD and same audience should be used in ruby code.

**Pull request closed without merging**